### DISCUSSION TITLE
[General] Paymaster - How to send all the refunded ETH back to the user

### DISCUSSION THREAD

**Discussion by lutr0** [zkSync-Community-Hub/zksync-developers#278] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I would like to implement a simple paymaster flow in `postTransaction` that sends all the the ETH refunded back to the user.
My expectation was that `_maxRefundedGas` covers `uint256 requiredETH = _transaction.gasLimit *_transaction.maxFeePerGas;`,

instead from my experimentation below it seems to give only refer to the gasLimit used.

So my questions are:

1.  What is the recommended wat to send all the the ETH refunded back to the user?
2.  Is recommended to set maxFeePerGas = gasPrice?

### What is the specific issue or error you're encountering?

Sending back only `_maxRefundedGas` 

```sol
function postTransaction(
        bytes calldata _context,
        Transaction calldata _transaction,
        bytes32,
        bytes32,
        ExecutionResult _txResult,
        uint256 _maxRefundedGas
    ) external payable override onlyBootloader {
        console.log("maxRefundedGas: ", _maxRefundedGas);
        address userAddress = address(uint160(_transaction.from));
        uint gasUsed = _txResult.gasUsed;
        (bool success, ) = userAddress.call{value: _maxRefundedGas}("");
        require(success, "Failed to refund after the transaction.");
    }
```
has the user refunded `0.000000000002682814` ETH 
while the paymaster received back `0.0006881845` ETH (see below)
![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/76906369/0a94f25e-af3d-4562-ab43-153788ff7ae6)

Multiplying by `_transaction.maxFeePerGas` seems to give us a good approximation:
` (bool success, ) = userAddress.call{value: _maxRefundedGas * _transaction.maxFeePerGas}("");`

User refunded `0.0006706995` ETH, paymaster refunded `0.00068476375`

![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/76906369/42ddff2e-86aa-4c8d-801b-bcc34703bac1)

However, it requires to use the current gasPrice as maxFeePerGas, which could lead to failed transactions if the gas spikes.

### Can you share the error messages or logs you're receiving, if any?

N/A

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

N/A

### Can you provide the relevant portions of your contract code where the issue is occurring?

N/A

### Have you tried to isolate the problem, and if so, what were the results?

N/A

### What steps have you already taken to try to resolve the issue?

N/A

**Comment by StanislavBreadless** [comment, marked as answer]
Hey! The `_maxRefundedGas` is the number of how much gas (not ETH) has been left after all the execution steps. 

It does not include:
- The fact that paymaster *itself* will consume a certain amount of gas in the postOp.
- Potential refunds that might come from places like a storage slot that has not been edited.

If you try to refund the user with anything above, then `_maxRefundedGas` there might be situations when the paymaster will not get compensated. 

A more correct way for implementing functionality like these is: to send something around `block.gasprice * (_maxRefundedGas - PAYMASTER_GAS` to the user), where `PAYMASTER_GAS` is some constant of how much gas a paymaster is expected to consume. However, we do not recommend relying on any sort of constants regarding gas costs as these might change during an upgrade (so to maintain this paymaster you'll have to carefully watch out for upgrades). 

Overall, while returning all the refunded eth to the user offers a better UX in some cases, there is no canonical easy way to do it

**Comment by discpenromo** [comment]
It seems that the issue lies in how the _maxRefundedGas value is calculated and used in the postTransaction function. Instead of relying solely on _maxRefundedGas, which may not accurately represent the refunded ETH, consider using a more precise calculation based on _transaction.gasLimit and _transaction.maxFeePerGas. Multiplying these values together should provide a better approximation of the refunded ETH amount. However, keep in mind the potential risk of using the current gas price as maxFeePerGas, as it may lead to failed transactions during gas spikes. Consider implementing safeguards or alternative approaches to handle this scenario.

**Comment by maggielyon** [comment]
To ensure that all refunded ETH is sent back to the user, it's recommended to multiply `_maxRefundedGas` by `_transaction.maxFeePerGas` when sending the refund. However, using the current gas price as `maxFeePerGas` could lead to failed transactions during gas spikes. Consider implementing a dynamic strategy to handle gas prices or providing users with the option to manually adjust gas fees.

**Comment by educatedb** [comment]
To send all refunded ETH back to the user, multiply _maxRefundedGas by _transaction.maxFeePerGas. However, be cautious of potential transaction failures due to gas spikes when setting maxFeePerGas to gasPrice.

**Comment by mancinottimoe** [comment]
To ensure that all refunded ETH is sent back to the user, you can use the formula:
(bool success, ) = userAddress.call{value: _maxRefundedGas * _transaction.maxFeePerGas}("");
This method may require setting maxFeePerGas to the current gasPrice, but it ensures accurate refunding without relying solely on _maxRefundedGas.
